### PR TITLE
Issue #2646 - Better handle concurrent calls to change session id and invalidate within a context 

### PR DESCRIPTION
**Pull request by janbartel**

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java [hunk 1]
```diff
@@ -813,19 +823,66 @@ public void renewId(HttpServletRequest request)
         String extendedId = null;
         try (Lock lock = _lock.lock())
         {
-            checkValidForWrite(); //don't renew id on a session that is not valid
+            while (true)
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java]: This loops seems superfluous.  The switch statement will just spin if the _state isn't one of the cases declared.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java [hunk 2]
```diff
@@ -350,6 +350,60 @@ public void clearEventListeners()
...
+    {
+        if (session == null)
+            return;
+        
+        if (_sessionListeners!=null)
+        {
+            HttpSessionEvent event=new HttpSessionEvent(session);      
+            for (int i = _sessionListeners.size()-1; i>=0; i--)
+            {
+                _sessionListeners.get(i).sessionDestroyed(event);
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: If more then 1 listener, and the first one fails, the others will not be called.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java [hunk 3]
```diff
@@ -350,6 +350,60 @@ public void clearEventListeners()
...
+    {
+        if (session == null)
+            return;
+        
+        if (_sessionListeners!=null)
+        {
+            HttpSessionEvent event=new HttpSessionEvent(session);      
+            for (int i = _sessionListeners.size()-1; i>=0; i--)
+            {
+                _sessionListeners.get(i).sessionCreated(event);
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: If more then 1 listener, and the first one fails, the others will not be called.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java [hunk 4]
```diff
@@ -350,6 +350,60 @@ public void clearEventListeners()
...
+    
+    protected void callSessionIdListeners (Session session, String oldId)
+    {
+        //inform the listeners
+        if (!_sessionIdListeners.isEmpty())
+        {
+            HttpSessionEvent event = new HttpSessionEvent(session);
+            for (HttpSessionIdListener l:_sessionIdListeners)
+            {
+                l.sessionIdChanged(event, oldId);
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: If more then 1 listener, and the first one fails, the others will not be called.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java (hunk 1 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java]: Added default clause to throw IllegalStateException

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java (hunk 2 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: Yes, I think that's reasonable. The listeners do not throw any checked exceptions. If one was to throw an unchecked exception, what's the most reasonable thing to do: keep on calling more listeners, gathering even more potential exceptions, and then throw them as unchecked exceptions too?

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java (hunk 3 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: See above

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java (hunk 4 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java]: see above

**Pull request merged**